### DISCUSSION TITLE
Handling External Repositories and Custom Changes

### DISCUSSION THREAD

**Discussion by KAGEYAM4** [twpayne/chezmoi#3393] [Q&A]
I utilize a variety of external repositories from where i get dotfiles/configs for applications, such as:

- https://github.com/jonniek/mpv-playlistmanager
- https://github.com/CogentRedTester/mpv-file-browser
- https://github.com/po5/thumbfast
- https://github.com/Eisa01/mpv-scripts
- https://github.com/prasanthrangan/hyprdots
- https://github.com/CogentRedTester/mpv-scroll-list
- and many more

Many of these repositories receive frequent updates, and some require custom changes made by the user. Currently I ->
- manually keep track of updates
- download them
- diff the changes
- determine what modifications were made by me versus the developer.
- apply changes as needed
- repeat

At this point i have given up and don't even take look at new configs of these repos.

--------------------------------------------------------------
I'm wondering if chezmoi could simplify this process. In the documentation, I came across https://www.chezmoi.io/user-guide/include-files-from-elsewhere/. Is this feature suitable for what I'm looking for? As a newcomer to Linux and dotfiles, the documentation seemed a bit overwhelming, and I'd appreciate some clarification and suggestions.

**Comment by halostatue** [comment]
It depends, and I do not know enough about the various items you’re talking about to answer comprehensively, but the answer is *mostly yes*.

1. Chezmoi can apply externals with no problems. *Modifying* files within the applied externals is problematic and will not work the way you want it to.
2. Many externals sources have configuration or cache files that are not part of the external path but live within the external path (see #3267 for an example of the latter). That configuration file can be managed directly by chezmoi with the correct external and chezmoi setup.
   - Your first item talks about a config file in [Settings](https://github.com/jonniek/mpv-playlistmanager#settings).

I don't use many externals, those that I use don't have separate configuration I care about, I’m not on Linux, and I don’t use MPV…so aside from the general information, I can't really help further, but I do hope that this helps.

**Comment by Lockszmith-GH** [comment]
I have some experience with chezmoi externals in Linux.

Treat these externals as immutable, if you need to modify something, modify a copy of them.  
You can automate these modifications using `after_...` scripts.

Note, that managing a large amount of externals, or even a few that are large, can slow down chezmoi operations, as their state is locally cached, and the cache needs to be refreshed from time to time (chezmoi does it automatically).

I've came up with a layered approach, where I manage the extranls separatly from my regular config, but keep everything in the same repo.
See discussion [here](https://github.com/twpayne/chezmoi/discussions/2375#discussioncomment-3735286) and some more details [here](https://github.com/twpayne/chezmoi/discussions/2574)

**Reply by KAGEYAM4** [reply]
Thank you, i will try to replicate your setup.

**Comment by bradenhilton** [comment]
I use `.chezmoiexternal` for mpv scripts, so I can speak to that specific use case.

```console
❯ cat .local\share\chezmoi\home\private_dot_config\mpv\.chezmoiexternal.toml | chezmoi execute-template
[fonts]
include = ['fonts/**']
stripComponents = 1
type = 'archive'
url = 'https://github.com/tomasklaen/uosc/releases/download/5.1.1/uosc.zip'

[scripts]
include = ['scripts/**']
stripComponents = 1
type = 'archive'
url = 'https://github.com/tomasklaen/uosc/releases/download/5.1.1/uosc.zip'

['scripts/sub-select.lua']
refreshPeriod = '168h'
type = 'file'
url = 'https://github.com/CogentRedTester/mpv-sub-select/raw/master/sub-select.lua'

['scripts/thumbfast.lua']
refreshPeriod = '168h'
type = 'file'
url = 'https://github.com/po5/thumbfast/raw/master/thumbfast.lua'
```

Here is the template that generates the external file:

```text
{{- $externals := dict -}}

{{- with $repo := "po5/thumbfast" -}}
{{-   $_ := set $externals "scripts/thumbfast.lua" (dict
        "type" "file"
        "url" (printf "https://github.com/%s/raw/master/thumbfast.lua" $repo)
        "refreshPeriod" "168h")
-}}
{{- end -}}

{{- with $repo := "tomasklaen/uosc" -}}
{{-   with $version := (gitHubLatestRelease $repo).GetTagName -}}
{{-     $_ := set $externals "fonts" (dict
          "type" "archive"
          "url" (printf "https://github.com/%s/releases/download/%s/uosc.zip" $repo $version)
          "stripComponents" 1
          "include" (list "fonts/**"))
-}}
{{-     $_ := set $externals "scripts" (dict
          "type" "archive"
          "url" (printf "https://github.com/%s/releases/download/%s/uosc.zip" $repo $version)
          "stripComponents" 1
          "include" (list "scripts/**"))
-}}
{{-   end -}}
{{- end -}}

{{- with $repo := "CogentRedTester/mpv-sub-select" -}}
{{-   $_ := set $externals "scripts/sub-select.lua" (dict
        "type" "file"
        "url" (printf "https://github.com/%s/raw/master/sub-select.lua" $repo)
        "refreshPeriod" "168h")
-}}
{{- end -}}

{{- $externals | toToml -}}
```
